planner/core/task.go

```diff
@@ -955,14 +955,36 @@ func (p *PhysicalTopN) getPushedDownTopN(childPlan PhysicalPlan) *PhysicalTopN {
 	return topN
 }
 
+// canPushToIndexPlan checks if this TopN can be pushed to the index side of copTask.
+// It can be pushed to the index side when all columns used by ByItems are available from the index side and
+//   there's no prefix index column.
+func (p *PhysicalTopN) canPushToIndexPlan(indexPlan PhysicalPlan, byItemCols []*expression.Column) bool {
+	schema := indexPlan.Schema()
+	for _, col := range byItemCols {
+		pos := schema.ColumnIndex(col)
```

Comment:
Do we also need to modify ColumnIndex as #22980 did?

Reply:
I think we don't need to.
This PR aims to fix #29711. Release-4.0 is a rather stable branch now, I prefer not to introduce unnecessary code. We can cherry-pick the `ColumnIndex` part if we need to cherry-pick #22980 to release-4.0.